_posts/2016-01-18-isolating-integration-tests-with-the-spring-framework.md

```diff
@@ -0,0 +1,141 @@
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    return "Delayed Result";
+  }
+
+}
+{% endhighlight %}
+
+What is the `@Profile` annotation used for? If we inject this controller into our standard application context, this has several drawbacks:
+
+ * The test will be slow: we only need to start one controller, not the whole thing
+ * Our controller will be picked up by Spring and injected into every other integration test, slowing down each integration test and maybe stepping on another test's toes
+
+A better alternative would be to spin up a minimal Spring Boot application exposing only our `DelayedWebServerController`. We will also tell Spring Boot to scan only the packages we are interested in, and to exclude persistence-related auto-configuration since we do not need it to spin up a controller. This is done in a `Configuration` class like this one:
+
+{% highlight java %}
+@Profile("restTemplateTimeout")
+@Configuration
+@EnableAutoConfiguration(
+    exclude = {DataSourceAutoConfiguration.class, HibernateJpaAutoConfiguration.class})
+@ComponentScan(basePackages = "my.application.resttemplate.timeout")
+public class DelayedWebServerConfiguration {
+    //The class is empty and only used to support the annotations
+}
+{% endhighlight %}
+
+The Spring context configuration can get quite confusing, let us look at the annotations one after the other:
+
+ * `@Profile`: This tells Spring that this configuration should only be used then the `restTemplateTimeout` profile is active. Further in this article, we will see how we enable this profile for a specific integration test. It is this annotation that prevents the configuration to be picked up by other unrelated integration tests. Note that our `DelayedWebServerController` is identically annotated.
+ * `@Configuration`: Standard annotation to tell Spring that this is a context configuration class.
+ * `@EnableAutoConfiguration`: Here we disable some of the Spring Boot "magic" that we do not need for our specific test
+ * `@ComponentScan`: We speed up the Spring Boot application startup by only scanning one package instead of the whole project. Any Spring-annotated class that is outside of this package will not be picked up by Spring.
+
+Here is how the integration test looks like:
+
+{% highlight java %}
+@RunWith(SpringJUnit4ClassRunner.class)
+@WebIntegrationTest("server.port:0")
+@SpringApplicationConfiguration(classes = DelayedWebServerConfiguration.class)
+@ActiveProfiles("restTemplateTimeout")
+public class RestTemplateShould {
+
+  @Rule
+  public ExpectedException thrown = none();
+
+  @Value("${local.server.port}")
+  private int port;
+
+  @Autowired
+  private RestTemplate restTemplate;
+
+  @Test
+  public void throw_timeout_if_response_lasts_more_than_two_seconds() {
+    thrown.expect(ResourceAccessException.class);
+    thrown.expectCause(instanceOf(SocketTimeoutException.class));
+
+    callEndpointWithDelay(3000);
```

Comment:
I don't think you want to use Awaitility in this case (it would call the `callEndpointWithDelay(3000)` every 100 milliseconds by default). Isn't the standard [JUnit timeout](https://github.com/junit-team/junit/wiki/Timeout-for-tests) mechanism what you're after?